### PR TITLE
n years ago and n months ago working

### PR DESCRIPTION
**Pull request by Saurabh3333**
Working for "n months ago" and "n years ago" inputs.

### CODE REVIEW

**Comment by Saurabh3333** [comment]
Why are the tests failing? I have manually checked for many inputs including "5 and 19 years ago" which is written in the test.js

**Comment by icyflame** [comment]
You have to remove `failing` from the test definition

**Comment by Saurabh3333** [comment]
Thanks, sir. Now it is ok.

File: index.js [hunk 1]
```diff
@@ -68,7 +68,27 @@ module.exports = function (str) {
   } else if ((match = str.match(new RegExp(regexps['re1'], 'i')))) {
     let days = match[1];
     let period = match[2];
-
+    if (period === 'year') {
+      let year = parseInt(new Date().getFullYear(), 10) - days;
+      if (year < 0) {
+        return 'Before B.C.';
+      } else {
+        return [true, new Date(new Date(new Date(new Date().setFullYear(year)).setMonth(new Date().getMonth())).setDate(new Date().getDate()))];
```
Comment (icyflame) [inline index.js]: You don't need to set current month. Date object returned by `new Date()` corresponds to present date.

File: index.js [hunk 2]
```diff
@@ -68,7 +68,27 @@ module.exports = function (str) {
   } else if ((match = str.match(new RegExp(regexps['re1'], 'i')))) {
     let days = match[1];
     let period = match[2];
-
+    if (period === 'year') {
+      let year = parseInt(new Date().getFullYear(), 10) - days;
```
Comment (icyflame) [inline index.js]: parse int is unnecessary

File: index.js [hunk 3]
```diff
@@ -68,7 +68,27 @@ module.exports = function (str) {
   } else if ((match = str.match(new RegExp(regexps['re1'], 'i')))) {
     let days = match[1];
     let period = match[2];
-
+    if (period === 'year') {
```
Comment (icyflame) [inline index.js:71]: does this handle both `1 year ago` and `n years ago`?

File: index.js [hunk 4]
```diff
@@ -68,7 +68,27 @@ module.exports = function (str) {
...
+        return 'Before B.C.';
+      } else {
+        return [true, new Date(new Date(new Date(new Date().setFullYear(year)).setMonth(new Date().getMonth())).setDate(new Date().getDate()))];
+      }
+    } else if (period === 'month') {
+      let extraYear = Math.floor(days / 12);
+      let extraMonth = days % 12;
+      let year = new Date().getFullYear() - extraYear;
+      let month = new Date().getMonth();
+      if (month >= extraMonth) {
```
Comment (icyflame) [inline index.js]: this can be done by subtracting and then taking a reminder (see the test for this case for a hint)

**Review by icyflame — CHANGES_REQUESTED** [review]
Thanks for your patch

File: index.js (hunk 3 above)
Comment (Saurabh3333) [reply index.js:71]: Yes. For every n.

File: index.js (hunk 4 above)
Comment (Saurabh3333) [reply index.js]: Thank You sir

File: index.js [hunk 5]
```diff
@@ -68,7 +68,26 @@ module.exports = function (str) {
...
+      let year = new Date().getFullYear() - extraYear;
+      var month = new Date().getMonth();
+      let extraMonth = (month - days) % 12;
+      if (extraMonth >= 0) {
+        month = extraMonth;
+      } else {
+        month = 12 + extraMonth;
+      }
 
+      return [true, new Date(new Date(new Date(new Date().setFullYear(year)).setMonth(month)))];
```
Comment (icyflame) [inline index.js]: Moving away from this format. use this https://github.com/icyflame/relative-date-reverse/commit/68757bc655f21c00a7b166f384dfaa084a252421

File: index.js [hunk 6]
```diff
@@ -68,7 +68,26 @@ module.exports = function (str) {
   } else if ((match = str.match(new RegExp(regexps['re1'], 'i')))) {
     let days = match[1];
     let period = match[2];
+    if (period === 'year') {
+      let year = new Date().getFullYear() - days;
+      if (year < 0) {
+        return 'Before B.C.';
+      } else {
+        return [true, new Date(new Date().setFullYear(year))];
```
Comment (icyflame) [inline index.js]: Moving away from this format. Use this instead https://github.com/icyflame/relative-date-reverse/commit/68757bc655f21c00a7b166f384dfaa084a252421

**Review by icyflame — CHANGES_REQUESTED** [review]

File: index.js [hunk 7]
```diff
@@ -68,7 +68,31 @@ module.exports = function (str) {
...
+        let updatedVal = new Date();
+        updatedVal.setFullYear(year);
+        return [true, updatedVal];
+      }
+    } else if (period === 'month') {
+      let extraYear = Math.floor(days / 12);
+      let year = new Date().getFullYear() - extraYear;
+      var month = new Date().getMonth();
+      let extraMonth = (month - days) % 12;
+      if (extraMonth >= 0) {
```
Comment (icyflame) [inline index.js]: The reminder of a division is never negative. This particular line isn't required. `extraMonth` is always `>= 0`

**Review by icyflame — CHANGES_REQUESTED** [review]

File: index.js (hunk 7 above)
Comment (Saurabh3333) [reply index.js]: But for the negative value of `extraMonth`, the `year` will also get reduced by one which is not true when `extraMonth` is positive.

Comment (icyflame) [reply index.js]: After line 84, `extraMonth` is positive. Your if/else is not required.

Comment (Saurabh3333) [reply index.js]: ((11-19)%12) = -8?

Comment (Saurabh3333) [reply index.js]: `month` will be between 0 to 11 but `days` can be anything?

Comment (Saurabh3333) [reply index.js]: sir, may I get some response?

Comment (icyflame) [reply index.js]: Please stop spamming maintainers. One message is enough.

Comment (icyflame) [reply index.js]: It looks like I was mistaken. Sorry about that.

What I can see is that, `x % y >= 0 if x > 0`

but `-1 * y < x % y <= 0` if `x < 0`.

**Review by icyflame — CHANGES_REQUESTED** [review]
Use something like this, I am no more comfortable with the "modular" approach if the reminder isn't always possible.

```js
let ey = Math.floor(days / 12)
let rem_months = days - ey * 12;
let cy = (new Date()).getFullYear()
cy -= ey;
let retVal = new Date()
retVal.setFullYear(cy);
retVal.setMonth(rem_months)
```

Node can handle negative values for setting months. i.e if you `setMonth(-4)`, Node will automatically set month to 8 ( ~ September)

See test for more details.

File: index.js [hunk 8]
```diff
@@ -68,7 +68,21 @@ module.exports = function (str) {
   } else if ((match = str.match(new RegExp(regexps['re1'], 'i')))) {
     let days = match[1];
     let period = match[2];
-
+    if (period === 'year') {
+      let year = new Date().getFullYear() - days;
+      if (year < 0) {
+        return 'Before B.C.';
```
Comment (icyflame) [inline index.js:74]: The return value should always be a `[<boolean>, <date-obj or null>]`

**Review by icyflame — CHANGES_REQUESTED** [review]
One final problem to fix

File: index.js (hunk 8 above)
Comment (Saurabh3333) [reply index.js:74]: For B.C what should I return: [true, null] or [false, null] or year in minus?

Comment (icyflame) [reply index.js:74]: I will take care of it.

**Review by icyflame — APPROVED** [review]

File: index.js (hunk 8 above)
Comment (Saurabh3333) [reply index.js:74]: Thank You

**Comment by icyflame** [comment]
Hey @Saurabh3333 , your code is merged into master. Thanks a lot for your work on this feature! :smile: :+1:

**Comment by Saurabh3333** [comment]
But it is not showing me, the code has been merged. It is showing `This branch is 10 commits ahead, 1 commit behind icyflame:master. `

**Comment by icyflame** [comment]
I have squashed all your commits into f7ef480 while merging.

**Comment by Saurabh3333** [comment]
Sir, I was asking because it is not reflecting my commits on KWoC statistics page which is important for final standing I guess.  Please look at it.

**Comment by icyflame** [comment]
This is some issue related to the KWoC dashboard, please contact people at KWoC for any resolution, I don't directly work on the stats page or anything related to it.

**Comment by icyflame** [comment]
@Saurabh3333 I have a clarification from Himanshu Mishra (https://github.com/orkohunter)

image

**Pull request closed without merging**